mypyc/irbuild/function.py

```diff
@@ -344,6 +353,20 @@ def c() -> None:
     return func_ir, func_reg
 
 
+def has_nested_func_self_reference(builder: IRBuilder, fitem: FuncItem) -> bool:
+    """Does a nested function contain a self-reference in its body?
+
+    If a nested function only has references in the surrounding function,
+    we don't need to add it to the environment.
+    """
+    if any(isinstance(sym, FuncBase) for sym in builder.free_variables.get(fitem, set())):
+        return True
+    return any(
+        has_nested_func_self_reference(builder, nested)
+        for nested in builder.encapsulating_funcs.get(fitem, [])
+    )
```

Comment:
I think it may be worth testing mutually recursive functions, something involving:
```python
def f():
    def g():
        h()
    def h():
        g()
    return g, h
```
or are we already testing this?

Reply:
Yeah, we already have at least the test case `testNestedFunctions2` in `mypyc/test-data/run-functions.test` (`mutual_recursion`).